algorithms/kernel/svm/svm_train_common_impl.i

```diff
@@ -0,0 +1,198 @@
+
+    /* Find i index of the working set (Bi) */
+    for (size_t i = 0; i < nActiveVectors; i++)
+    {
+        const algorithmFPType objFunc = grad[i];
+        if ((I[i] & up) && objFunc < GMin)
+        {
+            GMin = objFunc;
+            Bi   = i;
+        }
+    }
+    return GMin;
+}
+
+template <typename algorithmFPType, CpuType cpu>
+void HelperTrainSVM<algorithmFPType, cpu>::WSSjLocalBaseline(const size_t jStart, const size_t jEnd, const algorithmFPType * KiBlock,
+                                                             const algorithmFPType * kernelDiag, const algorithmFPType * grad, const char * I,
+                                                             const algorithmFPType GMin, const algorithmFPType Kii, const algorithmFPType tau,
+                                                             int & Bj, algorithmFPType & GMax, algorithmFPType & GMax2, algorithmFPType & delta)
+{
+    algorithmFPType fpMax = MaxVal<algorithmFPType>::get();
+    GMax                  = -fpMax; // some big negative number
+    GMax2                 = -fpMax;
+
+    const algorithmFPType zero(0.0);
+    const algorithmFPType two(2.0);
+
+    for (size_t j = jStart; j < jEnd; j++)
+    {
+        const algorithmFPType gradj = grad[j];
+        if ((I[j] & low) != low)
+        {
+            continue;
+        }
+        if (gradj > GMax2)
+        {
+            GMax2 = gradj;
+        }
+        if (gradj < GMin)
+        {
+            continue;
+        }
+
+        algorithmFPType b = GMin - gradj;
+        algorithmFPType a = Kii + kernelDiag[j] - two * KiBlock[j - jStart];
+        if (a <= zero)
+        {
+            a = tau;
+        }
+        algorithmFPType dt      = b / a;
+        algorithmFPType objFunc = b * dt;
+        if (objFunc > GMax)
+        {
+            GMax  = objFunc;
+            Bj    = j;
+            delta = -dt;
+        }
+    }
+}
+
+template <typename algorithmFPType, CpuType cpu>
+void HelperTrainSVM<algorithmFPType, cpu>::WSSjLocal(const size_t jStart, const size_t jEnd, const algorithmFPType * KiBlock,
+                                                     const algorithmFPType * kernelDiag, const algorithmFPType * grad, const char * I,
+                                                     const algorithmFPType GMin, const algorithmFPType Kii, const algorithmFPType tau, int & Bj,
+                                                     algorithmFPType & GMax, algorithmFPType & GMax2, algorithmFPType & delta)
+{
+    WSSjLocalBaseline(jStart, jEnd, KiBlock, kernelDiag, grad, I, GMin, Kii, tau, Bj, GMax, GMax2, delta);
+}
+
+template <typename algorithmFPType, CpuType cpu>
+services::Status SubDataTaskCSR<algorithmFPType, cpu>::copyDataByIndices(const uint32_t * wsIndices, const NumericTablePtr & xTable)
+{
+    DAAL_ITTNOTIFY_SCOPED_TASK(cache.copyDataByIndices.CSR);
+    CSRNumericTableIface * csrIface = dynamic_cast<CSRNumericTableIface *>(const_cast<NumericTable *>(xTable.get()));
```

Comment:
should this pointer be checked on nullptr?

Reply:
Applied